[must-gather] Add options for multi-node / host-network capture

alexanderConstantinescu opened the pull request:
This is must-gather feature enhancement coming from https://issues.redhat.com/browse/SDN-1760. 

In networking (possibly other components as well) we would need an ad-hoc ability to run must-gather commands against all/set of cluster nodes, as to facilitate customer/support in tasks that sometimes require several steps / manual interactions to perform. The most prevalent example of this is packet captures using `tcpdump`. Currently to run a packet capture on multiple nodes a customer/user needs to:

- Open X shell sessions
- In each: exec / run a pod with packet capture capabilities, specifying a file name and location for each
- Open a X+1 shell session which will be used to copy all `.pcap` files from the pods used in step 2, to the user's local location, by doing `oc cp` for each pod in step 2.  

Given that the [networking-dedicate debug image](https://github.com/openshift/network-tools/) already contains `tcpdump`: this process could be reduced to doing:

`oc adm must-gather --dest-dir ./tmp --source-dir '/tmp/tcpdump/' --image quay.io/openshift/origin-network-tools:latest --node-selector 'kubernetes.io/os=linux,node-role.kubernetes.io/master' --host -- timeout 30 tcpdump -i any -w /tmp/tcpdump/\$POD_NAME-%Y-%m-%dT%H:%M:%S.pcap -W 1 -G 300`

That single command would allow a user to run a packet capture on selected nodes (linux + master nodes) for 30 seconds and download all of the generated files. That is a significant improvement in usability from what currently exists.

soltysh commented on pkg/cli/admin/mustgather/mustgather.go:
```diff
@@ -626,6 +646,24 @@ func (o *MustGatherOptions) newPod(node, image string) *corev1.Pod {
+								},
+							},
+						},
+					},
```
What are these used for?

alexanderConstantinescu replied:
As mentioned in the commit message:

> This could be useful for users which specify `--node-selector` and capture must-gather data across multiple nodes, allowing them to separate the captured and downloaded data between nodes.

i.e: to have a separator for all downloaded files on a per node / pod basis.

soltysh commented on pkg/cli/admin/mustgather/mustgather.go:
```diff
@@ -87,6 +87,8 @@ func NewMustGatherCommand(f kcmdutil.Factory, streams genericclioptions.IOStream
 
 	cmd.Flags().StringVar(&o.NodeName, "node-name", o.NodeName, "Set a specific node to use - by default a random master will be used")
+	cmd.Flags().StringVar(&o.NodeSelector, "node-selector", o.NodeSelector, "Set a specific node selector to use - only relevant when specifying a specific command and image which needs to capture data on a set of cluster nodes simultaneously")
+	cmd.Flags().BoolVar(&o.HostNetwork, "host", o.HostNetwork, "Run must-gather pods as hostNetwork: true - relevant if a specific command and image needs to capture host-level data")
```
`host-network` let's make the flag explicit

alexanderConstantinescu replied:
Done

alexanderConstantinescu replied:
done

soltysh reviewed (changes requested):
Some changes and please squash your changes into single commit + eventually completions in a separate one.

alexanderConstantinescu commented on pkg/cli/admin/mustgather/mustgather.go:
```diff
@@ -321,13 +325,29 @@ func (o *MustGatherOptions) Run() error {
-		pod, err := o.Client.CoreV1().Pods(ns.Name).Create(context.TODO(), o.newPod(o.NodeName, image), metav1.CreateOptions{})
-		if err != nil {
-			return err
+		if o.NodeSelector != "" {
```
Done

alexanderConstantinescu replied:
done

alexanderConstantinescu commented:
> Some changes and please squash your changes into single commit + eventually completions in a separate one.

Excuse my delay on this. I completely lost track of the PR...I have addressed your comments and will keep a separate commit for that (what I assume you refer to as "completions") and squash all original commits into one.

npinaeva commented:
/retest-required

soltysh commented on pkg/cli/admin/mustgather/mustgather.go:
```diff
@@ -87,6 +87,8 @@ func NewMustGatherCommand(f kcmdutil.Factory, streams genericclioptions.IOStream
 	}
 
 	cmd.Flags().StringVar(&o.NodeName, "node-name", o.NodeName, "Set a specific node to use - by default a random master will be used")
+	cmd.Flags().StringVar(&o.NodeSelector, "node-selector", o.NodeSelector, "Set a specific node selector to use - only relevant when specifying a specific command and image which needs to capture data on a set of cluster nodes simultaneously")
```
Nit:
only relevant when specifying a ~~specific~~ command

soltysh commented on pkg/cli/admin/mustgather/mustgather.go:
```diff
@@ -248,6 +252,9 @@ func (o *MustGatherOptions) Validate() error {
 		return fmt.Errorf("missing an image")
 	}
+	if o.NodeName != "" && o.NodeSelector != "" {
+		return fmt.Errorf("\"node-name\" and \"node-selector\" are mutually exclusive: please specify one or the other")
```
Nit:
--node-name and --node-selector are mutually exclusive

soltysh reviewed (approved):
A few nits that can be addressed in a followup
/lgtm
/approve

soltysh commented:
/retest
